core/src/main/scala/akka/persistence/cassandra/journal/TagWriter.scala

```diff
@@ -0,0 +1,314 @@
+    case Flush =>
+      log.debug("Flushing")
+      if (buffer.nonEmpty) {
+        write(buffer, Vector.empty[Serialized], tagPidSequenceNrs)
+      }
+    case TagWrites(_, payload) =>
+      validate(payload)
+      flushIfRequired(buffer ++ payload, tagPidSequenceNrs)
+    case TagWriteDone(_) =>
+      log.error("Received Done when in idle state. This is a bug. Please report with DEBUG logs")
+  }
+
+  private def writeInProgress(buffer: Vector[Serialized], tagPidSequenceNrs: Map[String, Future[Long]]): Receive = {
+    case Flush =>
+    // Ignore, we will check when the write is done
+    case TagWrites(_, payload) =>
+      validate(payload)
+      // buffer until current query is finished
+      become(writeInProgress(buffer ++ payload, tagPidSequenceNrs))
+    case TagWriteDone(summary) =>
+      summary.foreach {
+        case (id, p @ PidProgress(_, seqNrTo, tagPidSequenceNr, offset)) =>
+          log.debug("Writing tag progress {}", p)
+          // These writes do not block future writes. We don't read he tag progress again from C*
+          // until a restart has happened. This is best effort and expect recovery to replay any
+          // events that aren't in the tag progress table
+          session.writeProgress(id, seqNrTo, tagPidSequenceNr, offset).onComplete {
+            case Success(_) =>
+              log.debug(
+                "Tag progress written. Pid: {} SeqNrTo: {} tagPidSequenceNr: {} offset: {}",
+                id, seqNrTo, tagPidSequenceNr, offset
+              )
+            case Failure(t) =>
+              log.error(t, s"Tag progress write has failed for pid: {} seqNrTo: {} tagPidSequenceNr: {} offset: {}. " +
+                s"If this is the only Cassandra error things will continue to work but if this keeps happening it till " +
+                " mean slower recovery as tag_views will need to be repaired.",
+                id, seqNrTo, tagPidSequenceNr, offset)
+          }
+      }
+      log.debug(s"Tag write complete. ${summary}")
+      flushIfRequired(buffer, tagPidSequenceNrs)
+      pubsub.foreach {
+        log.debug("Publishing tag update for {}", tag)
+        _.mediator ! DistributedPubSubMediator.Publish("akka.persistence.cassandra.journal.tag", tag)
+      }
+
+    case TagWriteFailed(t, events, previousSequenceNrs) =>
+      log.error(t, "Writing tags has failed. This means that any eventsByTag query will be out of date. The write will be retried.")
+      log.debug("Setting tag sequence nrs back to {}", previousSequenceNrs)
+      // wait for the next poll as not to hammer Cassandra
+      context.become(idle(events ++ buffer, previousSequenceNrs))
+  }
+
+  private def flushIfRequired(buffer: Vector[Serialized], tagSequenceNrs: Map[String, Future[Long]]): Unit = {
+    if (buffer.isEmpty) {
+      context.become(idle(buffer, tagSequenceNrs))
+    } else if (buffer.head.timeBucket < buffer.last.timeBucket) {
```

Comment:
timeBucket is derived from the time-uuid of the event? for a single persistenceId I understand that those are monotonically increasing, but do we know that for different persistenceId? where do we create that uuid? It must be created in the journal actor, before we delegate to futures (and loose the order)

Reply:
This happens during the serialise. Which I had moved to not be dispatched but then moved back to be dispatched when you mentioned that it'll improve serialisation.

Even if we create the UUIDs in order they get sent back the jounal actor after a cassandra write so they could then come to the TagWriter out of order.

I think it'll be best if the TagWriter doesn't assume they come in order across messages and sorts them. WDYT @patriknw ?

Reply:
right, but is it enough to sort? can't one batch be flushed and then some event for previous bucket arrives

Reply:
Yes this could happen but I don't think it is a problem as we now only assume that events for a given persistenceId are delivered in timeuuid / sequence nr order. 

The TagWriter and EventsByTagStage no longer rely on the buckets being written in order which I'd first done but gave up after observing lack of isolation in a C* unlogged batch. 

The TagWriter will write in the old bucket when they arrive and the stage will go back and find them if it sees a gap in the sequence numbers. 

I think the only thing that is broken right now is the check `buffer.head.timeBucket < buffer.last.timeBucket` to see if it has events from different buckets are in the current buffer so we should the earlier bucket right away.